### PR TITLE
arrays-programs_involving_data_sequences-hard-q7

### PR DESCRIPTION
**Pull request by KaiKhademi**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”
---
## Website
- [ ] Is the question text on the Gamification document up to date? 
    > Make sure to include the sample output if there is one for the question

### CODE REVIEW

File: src/arrays/programs_involving_data_sequences/hard/q7/SquareUp.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
+package arrays.programs_involving_data_sequences.hard.q7;
+
+/*
+Write a Java program that calls a method called doubler that takes as input an integer array, and squares each element
```
Comment (SamamthaH) [inline src/arrays/programs_involving_data_sequences/hard/q7/SquareUp.java:4]: Change question text to:
"Write a Java program that asks the user to input 10 integers and stores them in an array. Then it should pass them to a method called doubler takes an input of an integer array and stores the square of each element in a new array which it then returns. After calling the method, the main method should print the contents of the new array separated by a space. Your method should be general enough to work with any size of array."

File: src/arrays/programs_involving_data_sequences/hard/q7/MainTest.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    public static int n = 10;
+
+    public Clause[] testSentence() {
+        TestOption.isInputTest = true;
+        TestOption.defaultInput = arrayToInput(generateRandomArray(-10, 10, n));
+        return new Clause[]{
+                new StringLiteral("Enter an array of " + n + " integers to be squared:"),
+                new Optional(new StringLiteral(" ")),
+                new NewLine(),
+                new StringLiteral("The squared array is:"),
```
Comment (SamamthaH) [inline src/arrays/programs_involving_data_sequences/hard/q7/MainTest.java:31]: Add an optional space before the NewLine

File: src/arrays/programs_involving_data_sequences/hard/q7/MainTest.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                new NewLine(),
+                new PlaceHolder(),
+                new Optional(new StringLiteral(" "))
```
Comment (SamamthaH) [inline src/arrays/programs_involving_data_sequences/hard/q7/MainTest.java:34]:
```suggestion
```

Comment (SamamthaH) [reply src/arrays/programs_involving_data_sequences/hard/q7/MainTest.java:34]: PlaceHolder will capture any remaining data so this will never be used.

File: src/arrays/programs_involving_data_sequences/hard/q7/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    public static int[] sq(int[] arr) {
+        int[] input = arr.clone();
+        for (int i = 0; i < input.length; i++)
+            input[i] = (int) Math.pow(input[i], 2);
+        return input;
+    }
+
+    static Stream<Arguments> mainInputProvider() {
+        return Stream.of(
+                Arguments.of(generateRandomArray(-10000, 10000, n)),
```
Comment (SamamthaH) [inline src/arrays/programs_involving_data_sequences/hard/q7/MainTest.java:51]: Vary the range of these arrays.

File: src/arrays/programs_involving_data_sequences/hard/q7/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("mainInputProvider")
+    void printCorrectOutput(int[] a) throws InvalidClauseException {
+        TestOption.incorrectStructureErrorMessage = "Your program does not print the correct sequence of squared array.";
+        runWithInput(arrayToInput(a), new Clause[]{
+                new StringLiteral(arrayToInput(sq(a)))
```
Comment (SamamthaH) [inline src/arrays/programs_involving_data_sequences/hard/q7/MainTest.java:68]: do you need to add the optional here instead?

File: src/arrays/programs_involving_data_sequences/hard/q7/MainTest.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        });
+    }
+
+    static Stream<Arguments> methodInputProvider() {
+        return Stream.of(
+                Arguments.of(generateRandomArray(-10000, 10000, 10000)),
```
Comment (SamamthaH) [inline src/arrays/programs_involving_data_sequences/hard/q7/MainTest.java:74]: There should be some hardcoded arrays in here, and the generated ones should have varying lengths and ranges.

File: src/arrays/programs_involving_data_sequences/hard/q7/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    @ParameterizedTest
+    @MethodSource("methodInputProvider")
+    void correctFibonacciMaker(int[] input) throws Throwable {
+        Object[][] arguments = {
+                {input, int[].class},
+        };
+        MethodTest m = new MethodTest(SquareUp.class, "doubler", arguments);
+        m.setIncorrectMethodStructureErrorMessage("Your doubler method does not return the correct array of squared integers.");
+        Object output = m.callMethod();
+        assertArrayEquals((int[]) output, sq(input));
```
Comment (SamamthaH) [inline src/arrays/programs_involving_data_sequences/hard/q7/MainTest.java:93]: Please calculate the expected output before you call the method in case the student does not create a new array and passes back the original one instead. Also, move the expected value before the output, use CustomAssertions, and make you incorrect structure error a test error.

**Review by SamamthaH — CHANGES_REQUESTED** [review]

File: src/arrays/programs_involving_data_sequences/hard/q7/MainTest.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+    public void runMain() {
+        SquareUp.main(new String[0]);
+    }
+
+    public static int[] sq(int[] arr) {
+        int[] input = arr.clone();
+        for (int i = 0; i < input.length; i++)
+            input[i] = (int) Math.pow(input[i], 2);
+        return input;
```
Comment (SamamthaH) [inline src/arrays/programs_involving_data_sequences/hard/q7/MainTest.java:47]: To change this to be different than the solution code, you can have input[i]*=input[i] instead of Math.pow

**Review by SamamthaH — CHANGES_REQUESTED** [review]

File: src/arrays/programs_involving_data_sequences/hard/q7/MainTest.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+                Arguments.of(replicateArray(10000, n))
+
+
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("mainInputProvider")
+    void printCorrectOutput(int[] a) throws InvalidClauseException {
+        TestOption.incorrectStructureErrorMessage = "Your program does not print the correct sequence of squared array.";
```
Comment (SamamthaH) [inline src/arrays/programs_involving_data_sequences/hard/q7/MainTest.java:67]:
```suggestion
        TestOption.incorrectStructureErrorMessage = "Your program does not correctly print the squared array.";
```

**Review by SamamthaH — CHANGES_REQUESTED** [review]

**Review by SamamthaH — APPROVED** [review]
